src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/jvmsizing/OldGenReclamationRca.java

```diff
@@ -0,0 +1,147 @@
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+public class OldGenReclamationRca extends Rca<ResourceFlowUnit<HotResourceSummary>> {
+
+  private static final long EVAL_INTERVAL_IN_S = 5;
+  private static final double DEFAULT_TARGET_UTILIZATION_AFTER_GC = 75.0d;
+  private static final long DEFAULT_RCA_EVALUATION_INTERVAL_IN_S = 60;
+  private static final long B_TO_MB = 1024 * 1024;
+
+  private final MinOldGenSlidingWindow minOldGenSlidingWindow;
+  private final SlidingWindow<SlidingWindowData> gcEventsSlidingWindow;
+
+  private Metric heapUsed;
+  private Metric gcEvent;
+  private Metric heapMax;
+  private HotResourceSummary prevSummary;
+  private ResourceContext prevContext;
+  private double targetHeapUtilizationAfterGc;
+  private long rcaEvaluationIntervalInS;
+  private long rcaPeriod;
+  private int samples;
+
+  public OldGenReclamationRca(final Metric heapUsed, final Metric heapMax, final Metric gcEvent) {
+    this(heapUsed, heapMax, gcEvent, DEFAULT_TARGET_UTILIZATION_AFTER_GC,
+        DEFAULT_RCA_EVALUATION_INTERVAL_IN_S);
+  }
+
+  public OldGenReclamationRca(final Metric heapUsed, final Metric heapMax, final Metric gcEvent,
+      final double targetHeapUtilizationAfterGc, final long rcaEvaluationIntervalInS) {
+    super(EVAL_INTERVAL_IN_S);
+    this.heapUsed = heapUsed;
+    this.gcEvent = gcEvent;
+    this.heapMax = heapMax;
+    this.targetHeapUtilizationAfterGc = targetHeapUtilizationAfterGc;
+    this.rcaEvaluationIntervalInS = rcaEvaluationIntervalInS;
+    this.rcaPeriod = rcaEvaluationIntervalInS / EVAL_INTERVAL_IN_S;
+    this.samples = 0;
+    this.minOldGenSlidingWindow = new MinOldGenSlidingWindow(1, TimeUnit.MINUTES);
+    this.gcEventsSlidingWindow = new SlidingWindow<>(1, TimeUnit.MINUTES);
+    this.prevContext = new ResourceContext(State.UNKNOWN);
+    this.prevSummary = null;
+  }
+
+  @Override
+  public void generateFlowUnitListFromWire(FlowUnitOperationArgWrapper args) {
+    throw new UnsupportedOperationException("generateFlowUnitListFromWire should not be called "
+        + "for node-local rca: " + args.getNode().name());
+  }
+
+  @Override
+  public ResourceFlowUnit<HotResourceSummary> operate() {
+    samples++;
+    double oldGenMax = getOldGenValueForMetric(heapMax);
+    double oldGenUsed = getOldGenValueForMetric(heapUsed);
+    double gcEvents = getGcEvents();
+    long currTime = System.currentTimeMillis();
+    minOldGenSlidingWindow.next(new SlidingWindowData(currTime, oldGenUsed));
+    gcEventsSlidingWindow.next(new SlidingWindowData(currTime, gcEvents));
+
+    if (samples == rcaPeriod) {
+      samples = 0;
+      double events = gcEventsSlidingWindow.readSum();
+      if (events >= 1) {
+        double threshold = targetHeapUtilizationAfterGc / 100d * oldGenMax;
+        HotResourceSummary summary = null;
+        ResourceContext context = null;
+        if (minOldGenSlidingWindow.readMin() > threshold) {
```

Comment:
The operate method will only be called from one thread, so next(E e) and readMin() always happen in the right order. Multiple threads don't concurrently call this method. Not sure where the thread safety issue is arising from. I don't think we are sharing any static state too in the min sliding window class. Where are you seeing such inconsistencies?

Reply:
If anyone else modifies minOldGenSlidingWindow, or it's backed by an implementation that removes entries on a schedule instead of a write, this may be inconsistent.

It isn't inconsistent right now, so I won't block on this, but it's a pretty simple fix.